### PR TITLE
Added wait methods

### PR DESCRIPTION
**Pull request by hriddhighosh4**
## Description

1. Added wait functionalty. #7
2. Fixed  Issue #12

Added customwait

## Types of Changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] I have added tests to cover my changes.
- [x] My change requires a change to the documentation.
- [x] I have Extended README or added documentation/comments where applicable
- [ ] Added copyright headers for new files from CONTRIBUTING.md
- [x] My code follows the code style of this project.

- [x] I have read the [contributing guidelines](CONTRIBUTING.md).
- [x] I have updated the [ChangeLog](CHANGELOG.md) following [Semantic Versioning](https://semver.org)

### CODE REVIEW

File: engine/selenium_helper.py [hunk 1]
```diff
@@ -423,3 +424,50 @@ def get_locator_info(**kwargs):
...
+    wait_result = False
+    if mode != "hardwait":
+        locator, locator_value = get_locator_info(**kwargs)
+        logger.info("I'll wait for an UI element!")
+        # try for 2 times
+
+        for i in range(2):
+               try:
+                   if mode == "visibility":
+                      WebDriverWait(web_driver, 10).until(
```
Comment (AlphaRahul16) [inline engine/selenium_helper.py:450]: @hriddhighosh4 Why timeout 10 is hardcoded?

File: daksha_know-how/CreateTest.md [hunk 2]
```diff
@@ -28,10 +28,16 @@ task:
...
+  - wait_for:
+      mode: visibility
+      xpath: //div[@class='headertoprightDealerName']
   - validate_ui_element:
       mode: equals
       xpath: //div[@class='headertoprightDealerName']
       value: myKaarma
+  - wait_for:
+      mode: hardwait
+      value: 20
```
Comment (AlphaRahul16) [inline daksha_know-how/CreateTest.md:40]: add example for invisibility

File: daksha_know-how/CreateTest.md [hunk 3]
```diff
@@ -28,10 +28,16 @@ task:
       value: '{{ password }}'
   - click_button:
       xpath: //*[@id='bchromedriverSignIn']
+  - wait_for:
+      mode: visibility
+      xpath: //div[@class='headertoprightDealerName']
```
Comment (AlphaRahul16) [inline daksha_know-how/CreateTest.md:33]: Add wait for element to be clickable, wait for text to be and use fluent wait internally

File: engine/selenium_helper.py [hunk 4]
```diff
@@ -423,3 +424,50 @@ def get_locator_info(**kwargs):
...
+    wait_result = False
+    if mode != "hardwait":
+        locator, locator_value = get_locator_info(**kwargs)
+        logger.info("I'll wait for an UI element!")
+        # try for 2 times
+
+        for i in range(5):
+               try:
+                   if mode == "visibility":
+                      WebDriverWait(web_driver, 15).until(
```
Comment (AlphaRahul16) [inline engine/selenium_helper.py:450]: @hriddhighosh4 Please use fluent wait in python like this `wait = WebDriverWait(driver, 10, poll_frequency=1, ignored_exceptions=[ElementNotVisibleException, ElementNotSelectableException])
`

File: engine/selenium_helper.py [hunk 5]
```diff
@@ -423,3 +424,50 @@ def get_locator_info(**kwargs):
...
+                      break
+               except Exception as e:
+                   logger.error("Attempt " + str(i) + " for waiting for "+ mode +" of "+ locator +" failed \n", exc_info=True)
+    else :
+        try:
+            value = kwargs['value']
+        except KeyError:
+            return False, "Ill formatted arguments, 'value' must be present in the list of args for mode : hardwait"
+        logger.info("I'll wait "+ str(value) + " seconds")
+        time.sleep(value)
```
[the remark this answers is not in the document]
Comment (AlphaRahul16) [reply engine/selenium_helper.py:470]: @hriddhighosh4 User RXPY dependency to add reactive waits

File: examples/test.yml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  - launch_browser
+  - open_url:
+      url: https://qa.mykaarma.com
+  - fill_data:
+      xpath: //*[@id='firstPhaseInputs']//input[@placeholder='Username']
+      value: new-user
+  - click_button:
+      id: bSignIn
+  - wait_for:
+      mode: invisibility
```
Comment (prakashiit) [inline examples/test.yml:17]: @hriddhighosh4 This example doesn't belong here. It has several issues ranging from using our internal IPs to even displaying our credentials. Please use a different example, and the filename also needs improvement (`test.yml` doesn't make sense).

Comment (prakashiit) [reply examples/test.yml:17]: You can take a look at the existing examples in the repo.

File: engine/selenium_helper.py [hunk 7]
```diff
@@ -423,3 +425,51 @@ def get_locator_info(**kwargs):
...
+        logger.info("I'll wait for an UI element!")
+        # try for 2 times
+        wait = WebDriverWait(web_driver, 10, poll_frequency=1,
+                             ignored_exceptions=[ElementNotVisibleException, ElementNotSelectableException])
+        for i in range(5):
+               try:
+                   if mode == "visibility":
+                      wait.until(
+                       EC.visibility_of_element_located((locator, locator_value))
+                   )
```
Comment (prakashiit) [inline engine/selenium_helper.py:454]: the indentation on these closing braces isn't right @hriddhighosh4 . (same at line 461)

**Review by AlphaRahul16 — APPROVED** [review]

File: engine/selenium_helper.py [hunk 8]
```diff
@@ -423,3 +425,52 @@ def get_locator_info(**kwargs):
...
+     :param kwargs: WebElement Description/wait-time Fetched From YAML
+     :returns: Status of execution and Failure String
+     :rtype: tuple
+    """
+    try:
+        mode = kwargs['mode']
+    except KeyError:
+        return False, "Ill formatted arguments, 'mode' must be present in the list of args"
+    wait_result = False
+    if mode != "hardwait":
```
Comment (prakashiit) [inline engine/selenium_helper.py:444]: @hriddhighosh4 What if user provides an invalid value for `mode`? (say 'disappear')

File: engine/selenium_helper.py (hunk 7 above)
Comment (AlphaRahul16) [reply engine/selenium_helper.py]: instead of hardwait -- sleep make more sense

**Review by prakashiit — APPROVED** [review]

**Review by AlphaRahul16 — APPROVED** [review]

**Review by Rishi16royy — APPROVED** [review]

**Pull request merged**